src/meili/map_matcher.cc

```diff
@@ -216,8 +217,9 @@ FindMatchResult(const MapMatching& mapmatching,
     const auto& next_state = mapmatching.state(next_stateid);
     // It must stay on the first edge of the route
     for (auto label = state.RouteBegin(next_state); label != state.RouteEnd(); label++) {
-      if (label->edgeid.Is_Valid()) {
-        edgeid = label->edgeid;
+      if (label->edgeid().Is_Valid()) {
```

Comment:
> I am not sure why the map matching logic requires the first label to have an invalid edge id (could it check if predecessor is invalid?).

map matching logic requires the first label to be dummy, and we need a flag to indicate that the label is dummy. The flag could be invalid edgeid, or invalid predecessor. Since the first label can have valid edgeid now, so we can check predecessor instead.

We need the dummy label to tell if the route between same location (the origin pathlocation and a destination pathlocation are at the same node) is empty or is not found. If the `RouteBegin() == RouteEnd()` then the route is not found, otherwise, there is a route, even it's an empty route (i.e. `RouteBegin()-> predecessor()` is invalid)